apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,285 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local local_conf         = require("apisix.core.config_local").local_conf()
+local http               = require("resty.http")
+local core               = require("apisix.core")
+local ipairs             = ipairs
+local tostring           = tostring
+local type               = type
+local math_random        = math.random
+local error              = error
+local ngx                = ngx
+local ngx_timer_at       = ngx.timer.at
+local ngx_timer_every    = ngx.timer.every
+local string             = string
+local table              = table
+local string_sub         = string.sub
+local str_byte           = string.byte
+local str_find           = core.string.find
+local log                = core.log
+
+local default_weight
+local applications
+local auth_path
+local service_list_path
+local instance_list_path
+
+local schema = {
+    type = "object",
+    properties = {
+        host = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+            },
+        },
+        fetch_interval = {type = "integer", minimum = 1, default = 30},
+        prefix = {type = "string", default = "/nacos/v1/"},
+        auth_path = {type = "string", default = "auth/login"},
+        service_list_path = {type = "string", default = "ns/service/list?pageNo=1&pageSize=20"},
```

Comment:
Only return the first 20 services is not a good idea.
We need to iterate the upstreams and build a service list from those that have `nacos` discovery_type.
Like this:
https://github.com/apache/apisix/blob/e9a59ac5e8618737f69eddbaf05097b7ffab46ff/apisix/control/v1.lua#L144-L149

You might need some lazy load trick to break the dependent cycle.

Reply:
fixed.